duo_client/client.py

```diff
@@ -221,12 +221,24 @@ def api_call(self, method, path, params):
         * params: dict mapping from parameter name to stringified value,
             or a dict to be converted to json.
         """
+        params_go_in_body = method in ('POST', 'PUT', 'PATCH')
         if self.sig_version in (1, 2):
             params = normalize_params(params)
-        elif self.sig_version in (3, 4):
+            # v1 and v2 canonicalization don't distinguish between
+            # params and body. There's no separate body input.
+            body = None
+        elif self.sig_version == 3:
             # Raises if params are not a dict that can be converted
             # to json.
             params = self.canon_json(params)
+            body = params
+        elif self.sig_version == 4:
+            if params_go_in_body:
+                body = self.canon_json(params)
+                params = {}
+            else:
+                body = ''
```

Comment:
Is the distinction between an empty string for body (here) vs. None (as at line 229) important?

Reply:
Line 229 is for canonicalization versions where the body must never be concatenated into the canonicalized representation of the request. I like setting body to a non-string value in those cases because future bugs that tried to use the body wrong would raise, making them easier to catch early.

Perhaps body should be a required parameter to canonicalize() -- or default to None -- for that reason?

Reply:
I wanted to make canonicalize backwards compatible, so I didn't make it required, but I will default to None.

Reply:
> Is the distinction between an empty string for body (here) vs. None (as at line 229) important?

For sig v4, the body must be an empty string for GET and a json string for other request methods. For other sig versions, it's ignored.